### ISSUE TITLE
conflict with vim-fugitive <enter> mapping

### ISSUE THREAD

**Issue by ri-aje** [sheerun/vim-polyglot#363]
`tpope/vim-fugitive` is able to navigate git repo with mapped <cr> key. I found the mapping is unavailable whenever I enabled `sheerun/vim-polyglot` together with fugitive. below is a min repro vimrc.
```
set nocompatible
call plug#begin('~/.vim/bundle')
 Plug 'tpope/vim-fugitive'
 Plug 'sheerun/vim-polyglot'
call plug#end()
```
take `vim-polyglot` as an example, if I do:
1. `cd vim-polyglot` and then `vim -u <vimrc-file-with-above-content>`
2. `:Gedit c161994e9607399a7b365ab274592bfc4f100306`, the top commit SHA as of this writing.
3. `:map <cr>` or `:map <enter>` shows no mapping found.
whereas if I comment out `Plug 'sheerun/vim-polyglot'` in the vimrc file, then in step 3. above, `:map <cr>` shows `n  <CR>        *@:<C-U>exe <SNR>25_GF("edit")<CR>` and <enter> navigation is available.

my questions:
1. can any one repro this with the above steps?
2. if so, is the fix on polyglot or fugitive? since fugitive works out of the box and polyglot seems to have damaged its functionality, I posted here first for possible solution.

**Comment by idbrii** [comment]
What behaviour do you expect to be mapped to `<CR>`? (I don't see anything in current fugitive help for what CR does in a fugitive-object.)

Fugitive never checks b:did_ftplugin and doesn't have a ftplugin/git.vim so I don't get how this is happening.

FYI, polyglot's git ftplugin is the same as the one distributed by vim. Vim's ftplugins are lower priority than user installed ones, so I guess that's why we don't see the issue without polyglot.

For me on tpope/vim-fugitive@02c26619c4d42d874216943eb9ad69cff371a109, doing `:Gedit HEAD~` also shows no fugitive command mapped to CR, but I didn't try removing polyglot or adding git to polyglot_disabled. Also, I have no expectations that CR would do anything.

**Comment by ri-aje** [comment]
> What behaviour do you expect to be mapped to `<CR>`? 

as I mentioned in the issue report, it should `n  <CR>        *@:<C-U>exe <SNR>25_GF("edit")<CR>`

> (I don't see anything in current fugitive help for what CR does in a fugitive-object.)

check out [this article](http://vimcasts.org/episodes/fugitive-vim-browsing-the-git-object-database/). search for `enter` in the page. or `:h fugitive_<CR>`

> 
> Fugitive never checks b:did_ftplugin and doesn't have a ftplugin/git.vim so I don't get how this is happening.
> 
> FYI, polyglot's git ftplugin is the same as the one distributed by vim. Vim's ftplugins are lower priority than user installed ones, so I guess that's why we don't see the issue without polyglot.

later I found a workaround. if I comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12), then the issue goes away. does that lead to any solution?

> 
> For me on [tpope/vim-fugitive@02c2661](https://github.com/tpope/vim-fugitive/commit/02c26619c4d42d874216943eb9ad69cff371a109), doing `:Gedit HEAD~` also shows no fugitive command mapped to CR, but **I didn't try removing polyglot**

you are not following the repro steps. try disabling polyglot then repeat your steps. it should now show you something like `n  <CR>        *@:<C-U>exe <SNR>106_GF("edit")<CR>`

> or adding git to polyglot_disabled. 

> Also, I have no expectations that CR would do anything.

please check out the article or vim help I mentioned above.

**Comment by idbrii** [comment]
> you are not following the repro steps. try disabling polyglot then repeat your steps. it should now show you something like `n <CR> *@:<C-U>exe <SNR>106_GF("edit")<CR>`

Nope. Neither removing polyglot nor stripping everything out doesn't show anything mapped to `<cr>`:

```
mv ~/.*vim* vimarchive/.
git clone vimarchive/.vim/bundle/fugitive ~/.vim
cp vimarchive/.vim/bundle/sensible/plugin/sensible.vim ~/.vimrc
```
(tpope/vim-sensible@7aa7dba2a2594c209d8063c4c1c56de130c362d3 and tpope/vim-fugitive@02c26619c4d42d874216943eb9ad69cff371a109)

When use your workaround, what do you get from `:verbose map <cr>`?

What version of fugitive are you using? Perhaps tpope removed this feature (that vimcast article is 8 years old)?

**Comment by ri-aje** [comment]
> > you are not following the repro steps. try disabling polyglot then repeat your steps. it should now show you something like `n <CR> *@:<C-U>exe <SNR>106_GF("edit")<CR>`
> 
> Nope. Neither removing polyglot nor stripping everything out doesn't show anything mapped to `<cr>`:

ok. then I am not sure why it doesn't repro on your end.

> 
> ```
> mv ~/.*vim* vimarchive/.
> git clone vimarchive/.vim/bundle/fugitive ~/.vim
> cp vimarchive/.vim/bundle/sensible/plugin/sensible.vim ~/.vimrc
> ```
> ([tpope/vim-sensible@7aa7dba](https://github.com/tpope/vim-sensible/commit/7aa7dba2a2594c209d8063c4c1c56de130c362d3) and [tpope/vim-fugitive@02c2661](https://github.com/tpope/vim-fugitive/commit/02c26619c4d42d874216943eb9ad69cff371a109))

I see you have sensible above. sanity check. did you follow my repro as is, or some equivalent form? trying to understand what's causing the discrepancy we see.

> 
> When use your workaround, what do you get from `:verbose map <cr>`?

I got this with my workaround or when only loading vim fugitive without polyglot.
```
n  <CR>        *@:<C-U>exe <SNR>27_GF("edit")<CR>
        Last set from ~/.vim/bundle/vim-fugitive/autoload/fugitive.vim
```

judging from the mapped command, I think it came from [this line](https://github.com/tpope/vim-fugitive/blame/a489c6e1d3fc265242fabfd171021b0ea02dacf8/autoload/fugitive.vim#L4035) in vim-fugitive. and tpope's commit message says `Map <CR> in blobs to jump to blamed commit`, so it's pretty obvious the enter key carries a designed feature.

> 
> What version of fugitive are you using? 

I am up-to-date on both repo.
```
vim-fugitive HEAD a489c6e1d3fc265242fabfd171021b0ea02dacf8
vim-polyglot HEAD c161994e9607399a7b365ab274592bfc4f100306
```
in case it matters, my vim is
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Apr 10 2018 21:31:58)
Included patches: 1-1453
```

> Perhaps tpope removed this feature (that vimcast article is 8 years old)?

I don't think so, the article is linked in https://github.com/tpope/vim-fugitive readme which is up-to-date. you may find it by searching `Browsing the git object database`. plus, `:h fugitive_<CR>` also hints the existence of the mapping.

**Comment by gmnimrod** [comment]
Hi,
I can reproduce this on neovim using the same revisions.

```
vim-fugitive HEAD a489c6e1d3fc265242fabfd171021b0ea02dacf8
vim-polyglot HEAD c161994e9607399a7b365ab274592bfc4f100306
NVIM v0.3.1
```

> later I found a workaround. if I comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12), then the issue goes away. does that lead to any solution?

This workaround works on my system as well.

> > > you are not following the repro steps. try disabling polyglot then repeat your steps. it should now show you something like `n <CR> *@:<C-U>exe <SNR>106_GF("edit")<CR>`
> > 
> > 
> > Nope. Neither removing polyglot nor stripping everything out doesn't show anything mapped to `<cr>`:
> 
> ok. then I am not sure why it doesn't repro on your end.

Just to make sure, did you run `:verbose map <cr>` after `Gedit`? The mapping is set only on git filetypes.

**Comment by ri-aje** [comment]
> Hi,
> I can reproduce this on neovim using the same revisions.
> 
> ```
> vim-fugitive HEAD a489c6e1d3fc265242fabfd171021b0ea02dacf8
> vim-polyglot HEAD c161994e9607399a7b365ab274592bfc4f100306
> NVIM v0.3.1
> ```
> > later I found a workaround. if I comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12), then the issue goes away. does that lead to any solution?
> 
> This workaround works on my system as well.
> 
> > > > you are not following the repro steps. try disabling polyglot then repeat your steps. it should now show you something like `n <CR> *@:<C-U>exe <SNR>106_GF("edit")<CR>`
> > > 
> > > 
> > > Nope. Neither removing polyglot nor stripping everything out doesn't show anything mapped to `<cr>`:
> > 
> > 
> > ok. then I am not sure why it doesn't repro on your end.
> 
> Just to make sure, did you run `:verbose map <cr>` after `Gedit`? The mapping is set only on git filetypes.

correct. I see that mapping after running `Gedit`, and to be clear, I don't see that mapping before running `Gedit`. as you found, it's only available on git filetypes which makes sense since `vim-polyglot/ftplugin/git.vim` should only be loaded for such buffers. I can see airline prints `[git]` in status line after running `Gedit`.

**Comment by idbrii** [comment]
> I see you have sensible above. sanity check. did you follow my repro as is, or some equivalent form? trying to understand what's causing the discrepancy we see.

No, I don't have Plug. But I get the same behaviour as you (no map) when both plugins installed, so I wasn't testing the plugin install scenario.

It looks like you're not removing your .vim files so it's possible something is leaking in there. (Example: if you had a ~/.vim/ftplugin everything in there would be loaded.) If you follow my steps it will give you a truly vanilla install (I use sensible for basic common settings like filetype on).

> judging from the mapped command, I think it came from [this line](https://github.com/tpope/vim-fugitive/blame/a489c6e1d3fc265242fabfd171021b0ea02dacf8/autoload/fugitive.vim#L4035) in vim-fugitive. and tpope's commit message says `Map <CR> in blobs to jump to blamed commit`, so it's pretty obvious the enter key carries a designed feature.

Try removing conditions around calls to `fugitive#MapJumps()`. If I do `:Gedit c161994e9607399a7b365ab274592bfc4f100306 | echo b:fugitive_type` I get `commit` not `blob` and some calls check for `blob`. Also, [these lines look related](https://github.com/tpope/vim-fugitive/blame/a489c6e1d3fc265242fabfd171021b0ea02dacf8/autoload/fugitive.vim#L1786-L1788). If installing polyglot changes when/how the filetype is detected then that could mess something up?

**Comment by ri-aje** [comment]
> > I see you have sensible above. sanity check. did you follow my repro as is, or some equivalent form? trying to understand what's causing the discrepancy we see.
> 
> No, I don't have Plug. But I get the same behaviour as you (no map) when both plugins installed, so I wasn't testing the plugin install scenario.
> 
> It looks like you're not removing your .vim files so it's possible something is leaking in there. (Example: if you had a ~/.vim/ftplugin everything in there would be loaded.) If you follow my steps it will give you a truly vanilla install (I use sensible for basic common settings like filetype on).

I just followed your steps (without the sensible part, so vim-fugitive is the only plugin). I then navigate to vim-polyglot repo and issued `vim` to open an instance. then `:Gedit c161994e9607399a7b365ab274592bfc4f100306` and I can still see `<cr>` is mapped to:
```
n  <CR>        *@:<C-U>exe <SNR>25_GF("edit")<CR>
        Last set from ~/.vim/autoload/fugitive.vim
```

to be extra certain, I made a new user with absolute minimum setting (install vim-plug and then paste the min repro vimrc on top of this thread) and observe the bug too.

> 
> > judging from the mapped command, I think it came from [this line](https://github.com/tpope/vim-fugitive/blame/a489c6e1d3fc265242fabfd171021b0ea02dacf8/autoload/fugitive.vim#L4035) in vim-fugitive. and tpope's commit message says `Map <CR> in blobs to jump to blamed commit`, so it's pretty obvious the enter key carries a designed feature.
> 
> Try removing conditions around calls to `fugitive#MapJumps()`. 

tried. it makes no difference. the only effective workaround is to comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12) as I found earlier.

> If I do `:Gedit c161994e9607399a7b365ab274592bfc4f100306 | echo b:fugitive_type` I get `commit` not `blob` and some calls check for `blob`. Also, [these lines look related](https://github.com/tpope/vim-fugitive/blame/a489c6e1d3fc265242fabfd171021b0ea02dacf8/autoload/fugitive.vim#L1786-L1788). 

with the vanilla setting suggested by your way, `:Gedit c161994e9607399a7b365ab274592bfc4f100306 | echo b:fugitive_type` also gives me `commit`, which makes sense in reference of `fugitive#MapJumps` definition since the `GF("edit")` appears in the `else` branch and we want control enters there in the case of `commit`.

> If installing polyglot changes when/how the filetype is detected then that could mess something up?

it seems like we are discussing whether `<cr>` is mapped by vim-fugitive to begin with, which I am sure it is since I have provided several evidence. @gmnimrod also confirmed the issue and my workaround from his end. but somehow, this bug just doesn't appear in your setting? maybe it's caused by the plugin managers we use since that appears the only meaningful difference we have. which plugin manager you use?

**Comment by idbrii** [comment]
I use pathogen. But in my test steps above cloning fugitive into ~/.vim is using no plugin manager.

> without the sensible part, so vim-fugitive is the only plugin

Doing that (no ~/.vimrc) may mean that basic settings like `filetype plugin indent on` aren't executed (depends on your version of vim and I'm not sure when defaults.vim was introduced). That's why I use sensible as a test vimrc. 

> tried. it makes no difference. the only effective workaround is to comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12) as I found earlier.

Commenting out that line didn't make `nmap <CR>` return anything for my normal config. Same for disabling git support from polyglot:

    let g:polyglot_disabled = get(g:, 'polyglot_disabled', [])
    call add(g:polyglot_disabled, 'git')

Executing `:edit` in the buffer does make `<CR>` mapped to the GF("edit") function like you've shown.

Can you remove polyglot from the equation? Try installing vim-git directly:

```
set nocompatible
call plug#begin('~/.vim/bundle')
 Plug 'tpope/vim-fugitive'
 Plug 'tpope/vim-git'
call plug#end()
```

If you can demonstrate the problem with that, then you can close this bug and rename your bug with tpope to be "CR doesn't work in Gedit commit buffer" or something.

**Comment by gmnimrod** [comment]
> Can you remove polyglot from the equation? Try installing vim-git directly:

@idbrii is right. With vim-git the bug is reproduced, without it fugitive works fine.

> Fugitive never checks b:did_ftplugin and doesn't have a ftplugin/git.vim so I don't get how this is happening.

> FYI, polyglot's git ftplugin is the same as the one distributed by vim. Vim's ftplugins are lower priority than user installed ones, so I guess that's why we don't see the issue without polyglot.

Turns out, the ftplugins are not the same. polyglot's and vim-git's versions are slightly newer than vim's and neovim's.

The bug is present only in the newer version of `ftplugin/git.vim`. When polyglot is not disabled via `g:polyglot_disabled`, it overrides the default ftplugin. It does it by setting `did_ftplugin`. That is the way recommended in the docs.
(`:h ftplugin-overrule` or `:h write-filetype-plugin`)
With the workaround suggested by @ri-aje, the default ftplugin is executed after the user plugin, undoing the bug.

[This commit](https://github.com/tpope/vim-git/commit/503d1a569e816b0b3770234aefff04b3eae26224) makes the difference between versions. The added regex causes the ftplugin not to set `b:git_dir` when the current directory is a fugitive URL (any path that starts with "fugitive://"). [Here, in fugitive's code](https://github.com/tpope/vim-fugitive/blob/a69b0b0fd6d88acb39433117f541fb8930287a86/plugin/fugitive.vim#L253:L257) fugitive checks for `b:git_dir` before defining its mapping. Since it is missing, it skips the mapping.
(`fugitive#MapJumps` is also called in a bunch of other places, but I find that this is where it is called, when polyglot/vim-git are not interfering).

In summary, is a vim-git and/or fugitive bug (until proven otherwise of course :wink:).

**Comment by ri-aje** [comment]
> I use pathogen. But in my test steps above cloning fugitive into ~/.vim is using no plugin manager.
> 
> > without the sensible part, so vim-fugitive is the only plugin
> 
> Doing that (no ~/.vimrc) may mean that basic settings like `filetype plugin indent on` aren't executed (depends on your version of vim and I'm not sure when defaults.vim was introduced). That's why I use sensible as a test vimrc.
> 
> > tried. it makes no difference. the only effective workaround is to comment out [this line](https://github.com/sheerun/vim-polyglot/blame/c161994e9607399a7b365ab274592bfc4f100306/ftplugin/git.vim#L12) as I found earlier.
> 
> Commenting out that line didn't make `nmap <CR>` return anything for my normal config. Same for disabling git support from polyglot:
> 
> ```
> let g:polyglot_disabled = get(g:, 'polyglot_disabled', [])
> call add(g:polyglot_disabled, 'git')
> ```
> Executing `:edit` in the buffer does make `<CR>` mapped to the GF("edit") function like you've shown.
> 
> Can you remove polyglot from the equation? Try installing vim-git directly:
> 
> ```
> set nocompatible
> call plug#begin('~/.vim/bundle')
>  Plug 'tpope/vim-fugitive'
>  Plug 'tpope/vim-git'
> call plug#end()
> ```
> If you can demonstrate the problem with that, then you can close this bug and rename your bug with tpope to be "CR doesn't work in Gedit commit buffer" or something.

hi, I just tired under a newly created user with min repro .vimrc. turns out the bug is not there this time, with either your above short vimrc or the one I posted originally. since vimplug just pulled all updates, this drove me to think some recent change fixed the bug. I looked at my plugins that received updates and noticed vim-fugitive has some new commits and binary searched its latest commits. I was able to identify the fix commit [666c4abc](https://github.com/tpope/vim-fugitive/commit/666c4abc1299ef2a8975f58e37534fafe599b670) from fugitive checked in today.

to be fair, I wasn't sure this has to be polyglot bug, and I made it clear in my original post. I also opened issue to vim-fugitive (https://github.com/tpope/vim-fugitive/issues/1174), but didn't receive any reply from tpope so far.

in any event, the bug is fixed. thanks for your time and patience.

**Comment by idbrii** [comment]
@ri-aje If it's fixed, do you want to close both bugs?
